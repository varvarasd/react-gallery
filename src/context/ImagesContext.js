import React, { useEffect, createContext, useReducer } from 'react';
import { getImages } from '../services';
import reducer from './reducer';

const ImagesContext = createContext();


const ImagesProvider = ({ children }) => {
    const initialState = { 
        images: [{ url: '', author: ''}], 
        index: 0, 
        isLoading: true 
    };
    const [state, dispatch] = useReducer(reducer, initialState);

    useEffect(() => {
        getImages(dispatch);
    }, []);

    const contextValues = {
        failedFetch: state.failedFetch,
        images: state.images || [],
        image: state.images && state.images[state.index],
        isLoading: state.isLoading, 
        dispatch
    };

    return <ImagesContext.Provider value={contextValues}>
        {children}
    </ImagesContext.Provider>
}

export { 
    ImagesContext as default,
    ImagesProvider
};