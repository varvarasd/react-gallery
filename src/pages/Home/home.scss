.container {
    display: flex;
    flex-direction: column;
    max-width: 1000px;
    margin: auto;
    align-items: center;
    justify-content: center;
    height: 100vh;
}

@media (max-width: 768px) {
    .container {
        justify-content: center;
    }
}