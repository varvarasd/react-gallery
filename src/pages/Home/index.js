import React, { useContext, Fragment } from 'react';
import {
    Layout,
    Button,
    Loading,
    Image,
    LeftArrow,
    RightArrow
} from '../../components';
import Error from '../Error';

import ImagesContext from '../../context/ImagesContext';
import './home.scss';

const Home = () => {
    const { images, image, isLoading, dispatch, failedFetch } = useContext(ImagesContext);

    if(!images.length || failedFetch) return <Error errorMessage={failedFetch}/>;

    const mainContent = <Fragment>
        <div className="content__wrapper">
            <Image url={image.url} alt={image.author} />
        </div>
        <div className="buttons__wrapper">
            <Button role="previous" addClass="previous" labelName="previous" action={() => dispatch({ type: 'PREVIOUS_IMAGE' })}>
                <LeftArrow width={"20px"}/>
            </Button>
            <p className="image__author">By {image && image.author}</p>
            <Button role="next"addClass="next" labelName="next" action={() => dispatch({ type: 'NEXT_IMAGE' })}>
                <RightArrow width={"20px"}/>
            </Button>
        </div>
    </Fragment>;

    const contentElement = isLoading ? <Loading /> : mainContent;

    const preloadImages = images.map(({url, author}) => <link key={author} priority={true} as="image" href={url} />);

    return (
        <Layout
            links={preloadImages}
            title='React Gallery - Home' 
            description="Simple react gallery - home page">
            <section className="container">
                {contentElement}
            </section>
        </Layout>
    );
}

export default Home;