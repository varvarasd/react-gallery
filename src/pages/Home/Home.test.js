import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Home from '.';
jest.mock('react-helmet-async');

import ImagesContext from '../../context/ImagesContext';

describe('Home', () => {
    const mockContextValues = {
        images: [{ url: 'image/url', author: 'some author'}],
        image: { url: 'current/image/url', author: 'some author current image'},
        isLoading: false,
        dispatch: jest.fn()
    };

    it('renders Home content if isLoading is false', () => {

        const { container, getByText } = render(
            <ImagesContext.Provider value={mockContextValues}>
                <Home />
            </ImagesContext.Provider>
            );
        expect(getByText(/By some author current image/)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('calls dispatch with NEXT_IMAGE', () => {
        const spy = jest.spyOn(mockContextValues, 'dispatch')

        const { getByRole } = render(
            <ImagesContext.Provider value={mockContextValues}>
                <Home />
            </ImagesContext.Provider>
            );
        const nextButton = getByRole(/next/);
        fireEvent.click(nextButton);
        expect(spy).toHaveBeenCalledWith({"type": "NEXT_IMAGE"});
    });

    it('calls dispatch with PREVIOUS_IMAGE', () => {
        const spy = jest.spyOn(mockContextValues, 'dispatch')

        const { getByRole } = render(
            <ImagesContext.Provider value={mockContextValues}>
                <Home />
            </ImagesContext.Provider>
            );
        const previousButton = getByRole(/previous/);
        fireEvent.click(previousButton);
        expect(spy).toHaveBeenCalledWith({"type": "PREVIOUS_IMAGE"});
    });

    it('renders loading component if isLoading is true', () => {
        mockContextValues.isLoading = true;

        const { getByText } = render(
            <ImagesContext.Provider value={mockContextValues}>
                <Home />
            </ImagesContext.Provider>
            );
        expect(getByText(/Loading/)).toBeInTheDocument()
    });
});