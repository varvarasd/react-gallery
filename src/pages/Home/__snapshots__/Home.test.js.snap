// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders Home content if isLoading is false 1`] = `
<div>
  <main
    class="main"
  >
    <section
      class="container"
    >
      <div
        class="content__wrapper"
      >
        <div
          class="image__wrapper"
        >
          <img
            alt="some author current image"
            class="image"
            src="current/image/url"
          />
        </div>
      </div>
      <div
        class="buttons__wrapper"
      >
        <button
          class="button previous"
          label="previous"
          role="previous"
        >
          <svg
            viewBox="0 0 50 80"
            width="20px"
          >
            <polyline
              fill="none"
              points="45.63,75.8 0.375,38.087 45.63,0.375 "
              stroke="#808080"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="5"
            />
          </svg>
        </button>
        <p
          class="image__author"
        >
          By 
          some author current image
        </p>
        <button
          class="button next"
          label="next"
          role="next"
        >
          <svg
            viewBox="0 0 50 80"
            width="20px"
          >
            <polyline
              fill="none"
              points="0.375,0.375 45.63,38.087 0.375,75.8 "
              stroke="#808080"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="5"
            />
          </svg>
        </button>
      </div>
    </section>
  </main>
</div>
`;
